### DISCUSSION TITLE
Unity Instantiate 3D MIDI bar Problem

### DISCUSSION THREAD

**Discussion by Aleng0618** [melanchall/drywetmidi#223] [Q&A]
Hi, I'm very sorry for this stupid question. I'm still a newb.

I want to Instantiate midi  bar ,but the midi bar of Instantiate is horizontal, not vertical, and I use a dictionary to find out whether the Key is equal to the midi Name, but the position of the serialized midi bar is wrong,Maybe I misunderstood.How can I modify it?

My code is as follows:

```csharp
    public void NoteCreat()
    {
        TempoMap tempoMap = midiFile.GetTempoMap();

        foreach (GameObject gameObject in KeyBoardKeys)  //88 Piano Key of 3D GameObject
        {
            foreach (Note a in midiFile.GetNotes())
            {
                var timeInSeconds = ((TimeSpan)a.TimeAs<MetricTimeSpan>(tempoMap)).TotalSeconds;
                var lengthInSeconds = ((TimeSpan)a.LengthAs<MetricTimeSpan>(tempoMap)).TotalSeconds;

                foreach (var dicKey in PianoKeyDetector.PianoNotes)
                {
                    if (a.UnderlyingNote.ToString() == dicKey.Key && dicKey.Value.name == gameObject.name)
                    {
                        Vector3 notePso = new Vector3(gameObject.transform.position.x, gameObject.transform.position.y + (a.Time / 1000), gameObject.transform.position.z);

                        GameObject SpwanObject = Instantiate(NotePrefabs, notePso, Quaternion.identity, noteManager.transform);
                        SpwanObject.transform.localScale = new Vector3(gameObject.transform.position.x * (float)lengthInSeconds, NotePrefabs.transform.position.y, 0.01f);
                        SpwanObject.transform.localRotation = Quaternion.Euler(0, 0, 90);
                    }
                }
            }
        }
    }
```

**Comment by melanchall** [comment]
@Aleng0618 I've updated your post with the code since it much easier to handle an issue when a problem and related code are placed together.

Well, in fact I don't understand where to look at :) I'm not a Unity programmer and don't work with 3D programming so I need your assistance to point me to problem lines. Also

> but the position of the serialized midi bar is wrong

I have no idea what do you mean. Please provide more details, describe your algorithm better, provide screenshots.

**Reply by Aleng0618** [reply]
@melanchall Hello, I have solved the midi bar Scale with midi note.length, but the midi position is still wrong.
Screenshots, code and midi files will be provided.

Here is the screenshot:
![SmallBee](https://user-images.githubusercontent.com/114123013/191891985-92e87937-6f41-45c7-94b4-fd00e74db9da.png)

Here is the midi file and code:(my code have mark)

[MIDI Code_and_midi.zip](https://github.com/melanchall/drywetmidi/files/9630711/MIDI.Code_and_midi.zip)

**Reply by melanchall** [reply]
Please explain what's wrong on the screenshot? What picture should be?

Just put yourself on my place. I don't know your task, your algorithm and so on :-)

**Reply by Aleng0618** [reply]
![圖片](https://user-images.githubusercontent.com/114123013/191894898-199b8c57-a754-4e5c-a925-9af7dbc380e5.png)
It should be like this that the MIDI Bar is in the correct position corresponding to the Keyboard key, but the MIDI Bar is not in the corresponding Key position.

**Reply by Aleng0618** [reply]
I want to do this
![圖片](https://user-images.githubusercontent.com/114123013/191895458-8290042e-6083-4c25-b056-27315761fd04.png)

**Reply by Aleng0618** [reply]
The algorithm is that the position of the generated 3D object is the time of the MIDI note, the Scale of the 3D object is the length of the MIDI note, and then the track corresponding to the Note, I use the dictionary string pair Note Name to determine which track the note should be on

**Reply by melanchall** [reply]
OK, thanks, looks clear now. To help you, I need to open Unity. Please provide me the entire folder of your Unity project.

Also I should note that I can look into it on next week only. I'm on vacation right now without access to my computer.

**Reply by Aleng0618** [reply]
OK, then can I send the unity zip file to your google mail m.hall.muz@gmail.com?
Sorry to interrupt your holiday, have a good time and see you next week.

**Reply by melanchall** [reply]
> OK, then can I send the unity zip file to your google mail?

Yes, please. 

> Sorry to interrupt your holiday, have a good time and see you next week.

Thanks! :-)

**Comment by Aleng0618** [comment]
@melanchall Hello, the file has been sent!
